crates/bevy_reflect/bevy_reflect_derive/src/container_attributes.rs

```diff
@@ -0,0 +1,234 @@
+use syn::punctuated::Punctuated;
+use syn::token::Comma;
+use syn::{Meta, NestedMeta, Path};
+
+// The "special" trait idents that are used internally for reflection.
+// Received via attributes like `#[reflect(PartialEq, Hash, ...)]`
+const PARTIAL_EQ_ATTR: &str = "PartialEq";
+const HASH_ATTR: &str = "Hash";
+const SERIALIZE_ATTR: &str = "Serialize";
+
+/// A marker for trait implementations registered via the `Reflect` derive macro.
+#[derive(Clone)]
+pub(crate) enum TraitImpl {
+    /// The trait is not registered as implemented.
+    NotImplemented,
+    /// The trait is registered as implemented.
+    Implemented,
+
+    // TODO: This can be made to use `ExprPath` instead of `Ident`, allowing for fully qualified paths to be used
+    /// The trait is registered with a custom function rather than an actual implementation.
+    Custom(Ident),
+}
+
+impl Default for TraitImpl {
+    fn default() -> Self {
+        Self::NotImplemented
+    }
+}
+
+/// A collection of traits that have been registered for a reflected type.
+///
+/// This keeps track of a few traits that are utilized internally for reflection
+/// (we'll call these traits _special traits_ within this context), but it
+/// will also keep track of all registered traits. Traits are registered as part of the
+/// `Reflect` derive macro using the helper attribute: `#[reflect(...)]`.
+///
+/// The list of special traits are as follows:
+/// * `Hash`
+/// * `PartialEq`
+/// * `Serialize`
+///
+/// When registering a trait, there are a few things to keep in mind:
+/// * Traits must have a valid `Reflect{}` struct in scope. For example, `Default`
+///   needs `bevy_reflect::prelude::ReflectDefault` in scope.
+/// * Traits must be single path identifiers. This means you _must_ use `Default`
+///   instead of `std::default::Default` (otherwise it will try to register `Reflectstd`!)
+/// * A custom function may be supplied in place of an actual implementation
+///   for the special traits (but still follows the same single-path identifier
+///   rules as normal).
+///
+/// # Example
+///
+/// Registering the `Default` implementation:
+///
+/// ```ignore
+/// // Import ReflectDefault so it's accessible by the derive macro
+/// use bevy_reflect::prelude::ReflectDefault.
+///
+/// #[derive(Reflect, Default)]
+/// #[reflect(Default)]
+/// struct Foo;
+/// ```
+///
+/// Registering the `Hash` implementation:
+///
+/// ```ignore
+/// // `Hash` is a "special trait" and does not need (nor have) a ReflectHash struct
+///
+/// #[derive(Reflect, Hash)]
+/// #[reflect(Hash)]
+/// struct Foo;
+/// ```
+///
+/// Registering the `Hash` implementation using a custom function:
+///
+/// ```ignore
+/// // This function acts as our `Hash` implementation and
+/// // corresponds to the `Reflect::reflect_hash` method.
+/// fn get_hash(foo: &Foo) -> Option<u64> {
+///   Some(123)
+/// }
+///
+/// #[derive(Reflect)]
+/// // Register the custom `Hash` function
+/// #[reflect(Hash(get_hash))]
```

Comment:
I'm wondering if it would make sense to add Default as a special trait- it definitely has special behavior, and allowing a special custom function for it like this would be nice. 
Outside of the scope of this PR, but occurred to me nonetheless.

Reply:
That makes sense to me. It would involve adding a `Reflect::reflect_default()` method or something though. "Special traits" aren't stored as type data like other traits (can't access it via the registry). So we either add that method to `Reflect` and remove `ReflectDefault` or we make all "special traits" accessible from the registry as well.

But yeah that's probably best for a follow up PR (maybe even in #4140).

Reply:
I like this in #4140 at first glance.